ssl/quic/quic_wire.c

```diff
@@ -0,0 +1,816 @@
+            || !WPACKET_quic_write_vlint(pkt, ossl_quic_vlint_encode_len(value))
+            || !WPACKET_quic_write_vlint(pkt, value))
+        return 0;
+
+    return 1;
+}
+
+/*
+ * QUIC Wire Format Decoding
+ * =========================
+ */
+int ossl_quic_wire_peek_frame_header(PACKET *pkt, uint64_t *type)
+{
+    return PACKET_peek_quic_vlint(pkt, type);
+}
+
+int ossl_quic_wire_skip_frame_header(PACKET *pkt, uint64_t *type)
+{
+    return PACKET_get_quic_vlint(pkt, type);
+}
+
+static int expect_frame_header_mask(PACKET *pkt,
+                                    uint64_t expected_frame_type,
+                                    uint64_t mask_bits,
+                                    uint64_t *actual_frame_type)
+{
+    uint64_t actual_frame_type_;
+
+    if (!ossl_quic_wire_skip_frame_header(pkt, &actual_frame_type_)
+            || (actual_frame_type_ & ~mask_bits) != expected_frame_type)
+        return 0;
+
+    if (actual_frame_type != NULL)
+        *actual_frame_type = actual_frame_type_;
+
+    return 1;
+}
+
+static int expect_frame_header(PACKET *pkt, uint64_t expected_frame_type)
+{
+    uint64_t actual_frame_type;
+
+    if (!ossl_quic_wire_skip_frame_header(pkt, &actual_frame_type)
+            || actual_frame_type != expected_frame_type)
+        return 0;
+
+    return 1;
+}
+
+int ossl_quic_wire_peek_frame_ack_num_ranges(const PACKET *orig_pkt,
+                                             uint64_t *total_ranges)
+{
+    PACKET pkt = *orig_pkt;
+    uint64_t ack_range_count;
+
+    if (!expect_frame_header_mask(&pkt, OSSL_QUIC_FRAME_TYPE_ACK_WITHOUT_ECN,
+                                  1, NULL)
+            || !PACKET_skip_quic_vlint(&pkt)
+            || !PACKET_skip_quic_vlint(&pkt)
+            || !PACKET_get_quic_vlint(&pkt, &ack_range_count))
+        return 0;
+
+    /* (cannot overflow because QUIC vlints can only encode up to 2**62-1) */
+    *total_ranges = ack_range_count + 1;
+    return 1;
+}
+
+int ossl_quic_wire_decode_frame_ack(PACKET *pkt,
+                                    uint32_t ack_delay_exponent,
+                                    OSSL_QUIC_FRAME_ACK *ack,
+                                    uint64_t *total_ranges) {
+    uint64_t frame_type, largest_ackd, ack_delay_raw;
+    uint64_t ack_range_count, first_ack_range, start, end, i;
+
+    /* This call matches both ACK_WITHOUT_ECN and ACK_WITH_ECN. */
+    if (!expect_frame_header_mask(pkt, OSSL_QUIC_FRAME_TYPE_ACK_WITHOUT_ECN,
+                                  1, &frame_type)
+            || !PACKET_get_quic_vlint(pkt, &largest_ackd)
+            || !PACKET_get_quic_vlint(pkt, &ack_delay_raw)
+            || !PACKET_get_quic_vlint(pkt, &ack_range_count)
+            || !PACKET_get_quic_vlint(pkt, &first_ack_range))
+        return 0;
+
+    if (first_ack_range > largest_ackd)
+        return 0;
+
+    start = largest_ackd - first_ack_range;
+
+    if (ack != NULL) {
+        ack->delay_time
+            = ossl_time_multiply(OSSL_TIME_US,
```

Comment:
This is confusing to me. I changed the definition of `ossl_time_multiply` so that the second argument is an integer to multiply by (notwithstanding that OSSL_TIME is actually an integer). OSSL_TIME_US is the OSSL_TIME-typed value, the delay value is the multiplier.

Reply:
I was under the impression that ack_delay_raw was an OSSL_TIME, if it's not all is good.

Reply:
I guess:

```c
ossl_time_multiply(OSSL_TIME_US, safe_mul_uint64_t(ack_delay_raw, 1UL << ack_delay_exponent, &err));
if (err)
    ...
```

would be the proper way to do this.